### ISSUE TITLE
Consider using a local DB to manage city name inconsistencies

### ISSUE THREAD

**Issue by MartinKolarik** [jsdelivr/globalping#163]
In some cases, different providers return different variants of the same city name. #162 adds basic normalization for "city" and "the" cases, but overall a more effective solution might need a local DB with a list of all known names used to resolve the correct name. A possible source of the data: https://www.kaggle.com/datasets/geonames/geonames-database?select=geonames.csv

This is not something to do right now but a possible future improvement.

**Comment by ayuhito** [comment]
Alternate datasets: http://www.geonames.org or https://www.getty.edu/research/tools/vocabularies/tgn/index.html

Getty has a much more permissive free license than Geonames. But Geonames is much more comprehensive (although I haven't fully verified the quality of the free license).

Both are actively maintained though.

**Comment by MartinKolarik** [comment]
The file I linked and essentially everything else I found was based on geonames, so yes, that's probably the ultimate source.

**Comment by ayuhito** [comment]
Note that since both sets use licenses that require attribution (CC 4.0 or ODC 1.0), we can use the README or personally I do like https://discord.com/licenses giving credit to everyone in general (and also helps partially audit the licenses we use). In this example, Discord attributes the caniuse's CC 4.0 license.

They also have a fun little popup if you scroll down to the OS libraries tab here: https://discord.com/open-source

**Comment by ayuhito** [comment]
I can give this a go if that's fine.

I was thinking I'll make a separate NPM package called normalise-location that updates itself every month with GH Actions and processes all the data to a usable format. That way we're not adding much more complexity here.

**Comment by jimaek** [comment]
Why would we need frequent updates? Cities rarely change names. We should use an existing source that will minimize the maintenance needed on our side

**Comment by ayuhito** [comment]
It'll be automated in the normalise-location repo and a "whatever" thing to update here like any other dep patch update. The data is fixed otherwise.

https://www.geonames.org/recent-changes.html - They do seem to regularly update their dataset daily (although almost all of it is not relevant), so I think a monthly automated GH Action to rerun the data processing scripts is all that's needed to maintain itself. If there's no change to the data relevant to us, it won't publish a new NPM version.

It's a little overkill but new towns and places still do get added infrequently. And since I'm going to have to make the scripts once anyways, adding an automated GH action to auto-update isn't much extra work.

**Comment by MartinKolarik** [comment]
@ayuhito let's try and see how it works.

**Comment by ayuhito** [comment]
TLDR; Using Geonames is too inconsistent unless we get one of their premium vetted datasets imo.

Parsing through the db and choosing the best names for it relies on a [bunch of patterns](https://github.com/ayuhito/normalise-location/blob/main/scripts/strip-alt.ts#L47). You then build the records and filter out any name conflicts [here](https://github.com/ayuhito/normalise-location/blob/main/scripts/strip-all.ts#L90).

The main problem lies in name conflicts within large countries and inconsistent records which makes this unreliable for prod use. For example, [Dublin](https://www.geonames.org/2964574/dublin.html) and [Dublin City](https://www.geonames.org/7778677/dublin-city.html). In this situation Dublin City takes precedence over Dublin as ADM > PPLC, but for some reason Dublin City does not have an English name failing tests. Flipping priority to PPLC could fix this, but breaks other tests, because other countries seem to have more info in their ADM records and not PPL records by pure chance. 

I've tried loads of combinations, smaller more filtered datasets, different precedences, etc. It hasn't been fruitful.

When a name conflict is found, a place has to take precedence. While I kept adding more test cases, I realised while it may be possible to iron it out for the more popular names, it's very easy for smaller places to get absorbed in the process should a larger place have a random alternate name e.g. New York has an alt name Chelsea, which is a valid location elsewhere. Or Staten Island would sometimes map to [Richmond](https://www.geonames.org/5139559/richmond-county.html) instead. While these were fixable, I cannot guarantee it's consistency for even smaller places and fixing some other bugs I'm aware of in the code will likely break these cases again. Bluntly speaking, I can get it mostly working, but I can never trust the results because it's guaranteed to be inconsistent in places and I think that's unsafe to use in prod.

It seemed [okay](https://github.com/ayuhito/normalise-location/runs/7597952810?check_suite_focus=true) at first, but adding [more](https://github.com/ayuhito/normalise-location/runs/7599557596?check_suite_focus=true) and [more](https://github.com/ayuhito/normalise-location/runs/7621398869?check_suite_focus=true) test cases showed new circular patterns where one fixes and breaks the other.

(You can ignore errors like Brussels since that's an easy to fix regex error).

Personally, I'm in favour of maintaining a manual map of names considering I don't think conflicts are that common and can be easily tracked. There's even less concern for conflicts if we asciify contents, so only the rarest of differences such as Geneve and Geneva can be manually mapped. The only other in our current demo that _could_ be mapped is Frankfurt am Main -> Frankfurt and that's not necessarily needed but just nitpicking.

**Comment by MartinKolarik** [comment]
> The main problem lies in name conflicts within large countries and inconsistent records which makes this unreliable for prod use. For example, [Dublin](https://www.geonames.org/2964574/dublin.html) and [Dublin City](https://www.geonames.org/7778677/dublin-city.html). In this situation Dublin City takes precedence over Dublin as ADM > PPLC, but for some reason Dublin City does not have an English name failing tests. Flipping priority to PPLC could fix this, but breaks other tests, because other countries seem to have more info in their ADM records and not PPL records by pure chance.

Here it seems reasonable to use the primary name if English alt. isn't available.

> New York has an alt name Chelsea, which is a valid location elsewhere.

Can you post a link to this record? I wanted to check how it looks but can't find this.

> Or Staten Island would sometimes map to [Richmond](https://www.geonames.org/5139559/richmond-county.html) instead.

Again can't find the records that would cause this, but also not sure if it's relevant - the city is still NYC, and we don't go down to city districts.

**Comment by ayuhito** [comment]
> Here it seems reasonable to use the primary name if English alt. isn't available.

It does seem reasonable, but not according to Geonames as doing so will break the example for [Ceske Budejovice](https://www.geonames.org/3077915/okres-ceske-budejovice.html) to Ceske Budejovice District. We could slap a bandaid to replace any terms including District but it doesn't feel like a solid solution imo. 

> Again can't find the records that would cause this, but also not sure if it's relevant - the city is still NYC, and we don't go down to city districts.

These records probably come from names being merged when one record takes precedence over another. I never documented why, but I recall that merging record names did fix specific tests. It's just one variable to experiment with.

This is the issue, we have too much useless data that interferes with what we actually want. E.g. some countries will class ADM2 as districts whilst others will mark them as legitimate cases we would want. 

A possible idea to experiment with is only using P records, or even filtering down to a smaller subset such as `PPLC, PPLA, PPL`. Conflicts are arising from duplicates of differing precedence, so this might reduce a lot of noise data sending us in wild directions.

But like I mentioned in #194, I don't think it's necessary to put more hours into something uncertain like this unless we're spending too much time on maintaining the manual map. The map is an iterative process that we can manually verify for the rare instances a naming issue occurs.

It took me 5 minutes to make that PR compared to the 10+ hours trying to make this work. Throwing untested names into a db like this isn't guaranteed to be safe no matter how much we tune it unless we test every name case which isn't gonna happen compared to the manual map. It's just not worth the opportunity cost compared to the extra 5 minutes we'll need adding a few more names to be mapped in the future.

**Comment by MartinKolarik** [comment]
Please give me clear examples/links to the geo records to understand in detail what causes the issues. So far the only thing I was able to verify is the missing EN alt name.

**Comment by ayuhito** [comment]
I'll have to run the variations later this week since I never documented the results.

**Comment by ayuhito** [comment]
### Test Data
- Base case from my last experiment: [Code](https://github.com/ayuhito/normalise-location/tree/7f77bad8a0c5732825d5c916fb2d92ca28a9cd61) + [Tests](https://github.com/ayuhito/normalise-location/runs/7843956167?check_suite_focus=true)
- Remove overwrite check for final names mapped to users leads to better results: [Commit](https://github.com/ayuhito/normalise-location/commit/a34d30965d4345da37786359ab4736be0f7d7672) + [Tests](https://github.com/ayuhito/normalise-location/runs/7843961881?check_suite_focus=true) - A proper fix to use the hierarchy could be added but it's a big pain of a refactor and I don't think this solution works anyways. St. Petersburg can be fixed through a special char regex change, so that can be ignored. Brussels is broken as it technically returns City of Brussels, which could be resolved adding another regex to filter 'city of', but it still is fundamentally broken when there is a valid Brussels record. 
- Merge the name records when an existing record is replaced with a higher weighted record: [Commit](https://github.com/ayuhito/normalise-location/commit/62ec6e7443a6d5e818ad185c11861bab686e72a1) + [Tests](https://github.com/ayuhito/normalise-location/runs/7844134482?check_suite_focus=true)
- Only have English records. No non-English names at all. - [Commit](https://github.com/ayuhito/normalise-location/commit/f7b27eedce8c9c5e2f5605a3f613183f1fc2dd7f) + [Tests](https://github.com/ayuhito/normalise-location/runs/7844468911?check_suite_focus=true) - Doesn't look problematic from tests as it changes things like Sao Paulo to San Paulo and that's fine. It's an issue for records that do not have any English names. It'll just pick the first record name instead which applies to smaller cities. The aforementioned Ceske Budejovice District error isn't present since it just is getting regexed out but would be present otherwise.
- Use more geocodes (adding PPL, PPLL, PPLX, PPLS, PPLG from around 150k geocodes to 550k) - [Commit](https://github.com/ayuhito/normalise-location/commit/c8184abc400ac053fa302de1f98184380e55db54) + [Diff broken so here's the line change](https://github.com/ayuhito/normalise-location/blob/main/src/types.ts#L2) + [Tests](https://github.com/ayuhito/normalise-location/runs/7844702299?check_suite_focus=true) - More room for error, fixes a bunch, breaks others. Differences like Petah Tikva and Tiqva are a non-issue.  
- No P records (except PPLC) - [Commit](https://github.com/ayuhito/normalise-location/commit/11e5a0ae13c9420bfa3b8a9bdf99719380e3e200) + [Diff broken here's line change](https://github.com/ayuhito/normalise-location/blob/main/src/types.ts#L4) + [Tests](https://github.com/ayuhito/normalise-location/runs/7844898886?check_suite_focus=true) - Fixes some, breaks others.

We can go on and there are many many variations to go around like A records only + no merging names. But there's no perfect variation. A proper fix for the big overwrite bug would resolve cases like Moscow and Москва, but you can see how chaotic and different the results are per iteration, so it's won't be a final solution.

**Comment by jimaek** [comment]
Where are we with this? What are we waiting to proceed with https://github.com/jsdelivr/globalping/pull/194 ?

**Comment by MartinKolarik** [comment]
Waiting on me to read the long comment above.

**Comment by alexey-yarmosh** [comment]
Reopening this as part of #264.

I've updated the @ayuhito implementation a bit. It is dealing only with EN name alternatives and only with P feature class, which seems valid to me as all providers will return EN variations of city names.

[Here](https://github.com/alexey-yarmosh/normalise-location/blob/main/tests/normalise-country.test.ts) the tests can be found. All of them are passing, and some of them are commented out with the reason why. Also, [alternateNamesCountry.json](https://github.com/alexey-yarmosh/normalise-location/blob/main/data/alternateNamesCountry.json) file is handy to review other cases.

I think that solves the issue.

**Comment by alexey-yarmosh** [comment]
City approximation by lat/long covers most of the city name inconsistencies across the providers. Currently there are only 2 of them for the connected probes:
"Santiago de Queretaro" vs "Queretaro"
"Nurnberg" vs "Nuremberg"

I will add both to the https://github.com/jsdelivr/globalping/blob/master/src/lib/geoip/altnames.ts and close the issue. If new inconsistencies arise issue can be reopened to apply complex city name normalisation logic.

**Comment by alexey-yarmosh** [comment]
Fixed as part of https://github.com/jsdelivr/globalping/pull/442
